Sources/TuistCore/Cache/CacheDirectoriesProvider.swift

```diff
@@ -0,0 +1,60 @@
+import Foundation
+import TSCBasic
+import TuistGraph
+import TuistSupport
+
+public protocol CacheDirectoriesProviding: AnyObject {
+    /// Returns the cache directory
+    var cacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where the build artifacts are cached.
+    var buildCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where hashes of modules that have been a part of successful test are cached
+    var testsCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where the projects generated for automation tasks are generated to
+    var projectsCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where the project description helper modules are cached.
+    var projectDescriptionHelpersCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where the project description helper modules are cached.
+    var manifestCacheDirectory: AbsolutePath { get }
+}
+
+public final class CacheDirectoriesProvider: CacheDirectoriesProviding {
+    public let cacheDirectory: AbsolutePath
+    private static let defaultDirectory = AbsolutePath(URL(fileURLWithPath: NSHomeDirectory()).path).appending(component: ".tuist")
+
+    public init(config: Config?) throws {
+        if let cacheDirectory = config?.cache?.path {
+            self.cacheDirectory = cacheDirectory
+        } else {
+            cacheDirectory = CacheDirectoriesProvider.defaultDirectory.appending(component: "Cache")
+        }
+        if !FileHandler.shared.exists(cacheDirectory) {
```

Comment:
Side effects in class constructors are usually not a great idea. As a developer consuming `CacheDirectoriesProvider`, I'd not expect IO operations to happen at initialization time. I'd rather remove the logic from here, and let the entity that needs to use that directory make sure it's in the right state.

Reply:
I agree @pepibumur. The code here was moved from Environment.swift (where the existence and setup happens on init). The init in Environment was not throwing because it was using a force unwrap, based on the comment of @fortmarek I change this one here for throwing. If we move this to the computed property we would either move back to use force unwrap or convert all the properties to throwing methods.
Should I still try to change in this PR?

Reply:
I think we can a) let a different component / user of this component make sure that `cacheDirectory` exists b) create a function along the lines of `setupCache()` that will create it.

Reply:
As commented offline with Marek, this was moved a level up in `CacheDirectoriesProviderFactory` so that it can create `CacheDirectoriesProvider` and make sure the directories are ready to be used or created otherwise :)